core/src/smartcontracts/isi/triggers.rs

```diff
@@ -1,22 +1,518 @@
+                        )
+                    }
+                    EventType::ExecuteTrigger => {
+                        self.by_call_triggers.remove(id).map(|_| ()).expect(
+                            "`Set::by_call_triggers` doesn't contain required id. This is a bug",
+                        )
+                    }
+                })
+                .is_some()
+        }
+
+        /// Check if [`Set`] contains `id`.
+        pub fn contains(&self, id: &TriggerId) -> bool {
+            self.ids.contains_key(id)
+        }
+
+        /// Modify repetitions of the hook identified by [`Id`].
+        ///
+        /// # Errors
+        ///
+        /// - If a trigger with the given id is not found.
+        /// - If updating the current trigger `repeats` causes an overflow. Indefinitely
+        /// repeating triggers and triggers set for exact time always cause an overflow.
+        pub fn mod_repeats(
+            &self,
+            id: &TriggerId,
+            f: impl Fn(u32) -> Result<u32, RepeatsOverflowError>,
+        ) -> Result<(), ModRepeatsError> {
+            let res = self
+                .inspect_by_id(id, |action| match action.repeats() {
+                    Repeats::Exactly(atomic) => {
+                        let new_repeats = f(atomic.get())?;
+                        atomic.set(new_repeats);
+                        Ok(())
+                    }
+                    _ => Err(ModRepeatsError::RepeatsOverflow(RepeatsOverflowError)),
+                })
+                .ok_or_else(|| ModRepeatsError::NotFound(id.clone()));
+            // .flatten() -- unstable
+            match res {
+                Ok(r) => r,
+                Err(e) => Err(e),
+            }
+        }
+
+        /// Handle [`DataEvent`].
+        ///
+        /// Finds all actions, that are triggered by `event` and stores them.
+        /// This actions will be inspected in the next [`Set::handle_data_event()`] call
+        // Passing by value to follow other `handle_` methods interface
+        #[allow(clippy::needless_pass_by_value)]
+        pub fn handle_data_event(&self, event: DataEvent) {
+            for entry in self.data_triggers.iter() {
+                let id = entry.key();
+
+                if id.domain_id.is_some() && id.domain_id.as_ref() != event.domain_id() {
+                    continue;
+                }
+
+                self.match_and_insert_trigger(event.clone(), entry.pair());
+            }
+        }
+
+        /// Handle [`ExecuteTriggerEvent`].
+        ///
+        /// Find all actions that are triggered by `event` and store them.
+        /// These actions are inspected in the next [`Set::inspect_matched()`] call.
+        pub fn handle_execute_trigger_event(&self, event: ExecuteTriggerEvent) {
+            let Some(entry) = self.by_call_triggers.get(&event.trigger_id) else {
+                return
+            };
+
+            self.match_and_insert_trigger(event, entry.pair());
```

Comment:
**Stylistic point (ignore at will)**

This could be rewritten with a `map_or`. An `if let` also conveys the control flow better in this case.